### PR TITLE
2019 04 in flight txn tracking clean

### PR DESCRIPTION
**Pull request by ariard**
#333 without WIP Bump Tx stuff

### CODE REVIEW

**Comment by TheBlueMatt** [comment]
From IRC:

<BlueMatt> ariard: wait, no, yea, I'm still confused, it looks like this acts purely based on height
<BlueMatt> ariard: what I thought it was going to do was something like: step (1) create transactions (ie what we do now), step (2) add some mapping from txid to things-this-tx-resolves, step (3) monitor for confirmation of any tacked txids, step (4) when you see one, kick off a block-count-timer and then once you get enough confirmations, handle it
<BlueMatt> it looks like #336 skips (2) and (3), though I dont think addint them would be too hard?

File: src/ln/channelmonitor.rs [hunk 1]
```diff
@@ -390,6 +393,9 @@ pub struct ChannelMonitor {
 
 	destination_script: Script,
 
+
+	onchain_events_waiting_threshold_conf: HashMap<u32, Vec<OnchainEvent>>,
```
[the remark this answers is not in the document]
Comment (ariard) [reply src/ln/channelmonitor.rs]: Better ?

File: src/ln/channelmonitor.rs [hunk 2]
```diff
@@ -340,6 +327,22 @@ struct LocalSignedTx {
 	htlc_outputs: Vec<(HTLCOutputInCommitment, Option<(Signature, Signature)>, Option<HTLCSource>)>,
 }
 
+/// Upon discovering of some classes of onchain tx by ChannelMonitor, we may have to take actions on it
+/// once they mature to enough confirmations (HTLC_FAIL_ANTI_REORG_DELAY)
+#[derive(Clone, PartialEq)]
+enum OnchainEvent {
+	/// Outpoint under claim process by our own tx, at maturation we remote it from our_claim_txn_waiting_first_conf
```
[the remark this answers is not in the document]
Comment (ariard) [reply src/ln/channelmonitor.rs]: Yes it's inside the added commit, what is best practice referring it even if not present in commit or doing a generic reference?

File: src/chain/chaininterface.rs [hunk 3]
```diff
@@ -78,7 +78,11 @@ pub trait ChainListener: Sync + Send {
 	fn block_connected(&self, header: &BlockHeader, height: u32, txn_matched: &[&Transaction], indexes_of_txn_matched: &[u32]);
 	/// Notifies a listener that a block was disconnected.
 	/// Unlike block_connected, this *must* never be called twice for the same disconnect event.
-	fn block_disconnected(&self, header: &BlockHeader);
+	///
+	/// Provide listeners with filtered txn previously registered as watched because channels are
```
[the remark this answers is not in the document]
Comment (ariard) [reply src/chain/chaininterface.rs]: Yes stale comment, removed!

File: src/ln/channelmonitor.rs [hunk 4]
```diff
@@ -401,6 +428,12 @@ pub struct ChannelMonitor {
 
 	destination_script: Script,
 
+	our_claim_txn_waiting_first_conf: HashMap<BitcoinOutPoint, (u32, (TxMaterial, u64), u64)>,
```
[the remark this answers is not in the document]
Comment (ariard) [reply src/ln/channelmonitor.rs]: Merged the amount with TxMaterial and commented, basically first u32 is a height timer and last u64 the last fee we used to broadcast transaction.

File: src/ln/channelmonitor.rs [hunk 5]
```diff
@@ -175,25 +175,7 @@ impl<Key : Send + cmp::Eq + hash::Hash> ChainListener for SimpleManyChannelMonit
 			for htlc in htlc_updated_infos.drain(..) {
 				match pending_htlc_updated.entry(htlc.2) {
 					hash_map::Entry::Occupied(mut e) => {
-						// In case of reorg we may have htlc outputs solved in a different way so
-						// we prefer to keep claims but don't store duplicate updates for a given
-						// (payment_hash, HTLCSource) pair.
-						// TODO: Note that we currently don't really use this as ChannelManager
```
[the remark this answers is not in the document]
Comment (ariard) [reply src/ln/channelmonitor.rs]: Ah yes, you're right. If we track right and don't broadcast htlc failures until HTLC_FAIL_ANTI_REORG_DELAY, there shouldn't be no more htlc failure duplicate, assuming reorgs don't get bigger than delay. Remove the TODO nevertheless as we delay backward failure directly into ChannelMonitor now

File: src/ln/channelmonitor.rs [hunk 6]
```diff
@@ -1216,16 +1314,29 @@ impl ChannelMonitor {
...
-									log_trace!(self, "Failing HTLC with payment_hash {} from {} remote commitment tx due to broadcast of revoked remote commitment transaction", log_bytes!(htlc.payment_hash.0), $commitment_tx);
-									htlc_updated.push(((**source).clone(), None, htlc.payment_hash.clone()));
+									log_trace!(self, "Failing HTLC with payment_hash {} from {} remote commitment tx due to broadcast of revoked remote commitment transaction, waiting confirmation until {} height", log_bytes!(htlc.payment_hash.0), $commitment_tx, height + HTLC_FAIL_ANTI_REORG_DELAY - 1);
+									match self.onchain_events_waiting_threshold_conf.entry(height + HTLC_FAIL_ANTI_REORG_DELAY - 1) {
+										hash_map::Entry::Occupied(mut entry) => {
+											let e = entry.get_mut();
+											e.retain(|ref event| {
+												match **event {
+													OnchainEvent::HTLCUpdate { ref htlc_update } => {
+														return htlc_update.0 != **source
```
[the remark this answers is not in the document]
Comment (ariard) [reply src/ln/channelmonitor.rs]: Hmmm, but you may have duplicates with block_connected getting called many times for same block?

File: src/chain/chaininterface.rs [hunk 7]
```diff
@@ -78,7 +78,7 @@ pub trait ChainListener: Sync + Send {
 	fn block_connected(&self, header: &BlockHeader, height: u32, txn_matched: &[&Transaction], indexes_of_txn_matched: &[u32]);
 	/// Notifies a listener that a block was disconnected.
 	/// Unlike block_connected, this *must* never be called twice for the same disconnect event.
-	fn block_disconnected(&self, header: &BlockHeader);
+	fn block_disconnected(&self, header: &BlockHeader, height: u32);
```
[the remark this answers is not in the document]
Comment (ariard) [reply src/chain/chaininterface.rs]: Done, and updated parameters name to disconnected_height

File: src/ln/channelmonitor.rs [hunk 8]
```diff
@@ -2046,7 +2046,17 @@ impl ChannelMonitor {
 					payment_preimage.0.copy_from_slice(&input.witness[1]);
 					htlc_updated.push((source, Some(payment_preimage), payment_hash));
 				} else {
-					htlc_updated.push((source, None, payment_hash));
+					log_trace!(self, "Failing HTLC with payment_hash {} timeout by a spend tx, waiting confirmation until {} height", log_bytes!(payment_hash.0), height + HTLC_FAIL_ANTI_REORG_DELAY - 1);
```
[the remark this answers is not in the document]
Comment (ariard) [reply src/ln/channelmonitor.rs]: Hmmm isn't more CLTV_EXPIRY_DELTA and HTLC_FAIL_ANTI_REORG_DELAY which are interwined ? Let's say remote commitment transaction hits onchain, and we resolves a received_HTLC with a timeout tx, after HTLC_FAIL_ANTI_REORG_DELAY we pass failure backward. As it's way inferior to CLTV_EXPIRY_DELTA, it should be safe, assuming would_broadcast_at_height react timely.

File: src/chain/chaininterface.rs [hunk 9]
```diff
@@ -279,11 +280,11 @@ impl ChainWatchInterfaceUtil {
 	}
 
 	/// Notify listeners that a block was disconnected.
-	pub fn block_disconnected(&self, header: &BlockHeader) {
+	pub fn block_disconnected(&self, block: &Block, disconnected_height: u32) {
```
[the remark this answers is not in the document]
Comment (ariard) [reply src/chain/chaininterface.rs]: Ah sorry, this is surely rest from previous design, where I thought txn in block were needed to cancel events. Corrected

File: src/ln/channelmonitor.rs [hunk 10]
```diff
@@ -1216,16 +1336,29 @@ impl ChannelMonitor {
...
-				// confirmed, but we also need to do more other tracking of in-flight pre-confirm
-				// on-chain claims, so we can do that at the same time.
 				macro_rules! check_htlc_fails {
 					($txid: expr, $commitment_tx: expr) => {
 						if let Some(ref outpoints) = self.remote_claimable_outpoints.get($txid) {
 							for &(ref htlc, ref source_option) in outpoints.iter() {
 								if let &Some(ref source) = source_option {
-									log_trace!(self, "Failing HTLC with payment_hash {} from {} remote commitment tx due to broadcast of revoked remote commitment transaction", log_bytes!(htlc.payment_hash.0), $commitment_tx);
-									htlc_updated.push(((**source).clone(), None, htlc.payment_hash.clone()));
+									log_trace!(self, "Failing HTLC with payment_hash {} from {} remote commitment tx due to broadcast of revoked remote commitment transaction, waiting confirmation until {} height", log_bytes!(htlc.payment_hash.0), $commitment_tx, height + HTLC_FAIL_ANTI_REORG_DELAY - 1);
```
[the remark this answers is not in the document]
Comment (ariard) [reply src/ln/channelmonitor.rs]: Done!

File: src/ln/channelmonitor.rs [hunk 11]
```diff
@@ -1663,48 +1819,93 @@ impl ChannelMonitor {
...
+				spendable_outputs.append(&mut $updates.1);
+				watch_outputs.append(&mut $updates.2);
+				for claim in $updates.3 {
+					self.our_claim_txn_waiting_first_conf.insert(claim.0, claim.1);
+				}
+			}
+		}
+
 		if let &Some(ref local_tx) = &self.current_local_signed_commitment_tx {
+			for &(ref htlc, _, ref source) in &local_tx.htlc_outputs {
```
[the remark this answers is not in the document]
Comment (ariard) [reply src/ln/channelmonitor.rs]: Oh yes modified to only fail backward if commitment tx is last or previous local one.

But we still need to iterate on both HTLCs set as you may receive a update_fail_htlc + commitment signed from your counterpart. In case of last local tx hitting the chain, our in-memory last local tx datas won't contain the failed dust HTLC output. That's a edge-case but test_failure_delay_dust_htlc_local_commitment is explicitly testing that.

File: src/ln/channelmonitor.rs [hunk 12]
```diff
@@ -1704,48 +1852,100 @@ impl ChannelMonitor {
...
+				spendable_outputs.append(&mut $updates.1);
+				watch_outputs.append(&mut $updates.2);
+				for claim in $updates.3 {
+					self.our_claim_txn_waiting_first_conf.insert(claim.0, claim.1);
+				}
+			}
+		}
+
+		// HTLCs set may differ between last and previous local commitment txn, in case of one hitting chain, ensure we cancel all of them backward
+		let is_local_tx = if self.current_local_signed_commitment_tx.is_some() || self.prev_local_signed_commitment_tx.is_some() { true } else { false };
```
Comment (TheBlueMatt) [inline src/ln/channelmonitor.rs]: This isnt enough to verify that the given transaction is a local transaction, is it? I think you *also* need to check that commitment_txid == local_tx.txid for either prev or current?

Comment (TheBlueMatt) [reply src/ln/channelmonitor.rs]: Probably worth trying to add a test case for this.

File: src/ln/channelmonitor.rs [hunk 13]
```diff
@@ -1780,7 +1980,7 @@ impl ChannelMonitor {
 			let mut res = vec![local_tx.tx.clone()];
 			match self.key_storage {
 				Storage::Local { ref delayed_payment_base_key, ref prev_latest_per_commitment_point, .. } => {
-					res.append(&mut self.broadcast_by_local_state(local_tx, prev_latest_per_commitment_point, &Some(*delayed_payment_base_key)).0);
+					res.append(&mut self.broadcast_by_local_state(local_tx, prev_latest_per_commitment_point, &Some(*delayed_payment_base_key), 0).0); // Accurate height doesn't matter as we're going to change
```
Comment (TheBlueMatt) [inline src/ln/channelmonitor.rs]: nit: excess newline at EOL.
Also, I found this comment very confusing. Maybe something more explicit like "We throw away the generated waiting_first_conf data as we aren't (yet) confirmed and we don't actually know what the caller wants to do. The data will be re-generated and tracked in check_spend_local_transaction if we get a confirmation."

File: src/ln/channelmonitor.rs (hunk 8 above)
[the remark this answers is not in the document]
Comment (TheBlueMatt) [reply src/ln/channelmonitor.rs]: Ah, yea, sorry, you're right, we should still try to check how far we are from where we need to have failed back by, but, indeed, its a question of CLTV_EXPIRY_DELTA.

File: src/ln/channelmanager.rs [hunk 14]
```diff
@@ -332,6 +332,8 @@ pub struct ChannelManager {
 	channel_state: Mutex<ChannelHolder>,
 	our_network_key: SecretKey,
 
+	channel_closing_waiting_threshold_conf: Mutex<HashMap<u32, Vec<[u8; 32]>>>,
```
[the remark this answers is not in the document]
Comment (ariard) [reply src/ln/channelmanager.rs]: Updated doc and moved it in ChannelHolder ? IIRC this buffer is just there to avoid inconsistencies between channel_manager and channel_monitor in case of reorgs and an HTLC-timeout being swept by a preimage tx, in this case we need to pass success backward and we need the channel to be still open. Do you think it's ok right now before we put more thoughts in channel locks refactoring ?

File: src/ln/channelmonitor.rs [hunk 15]
```diff
@@ -401,6 +445,15 @@ pub struct ChannelMonitor {
 
 	destination_script: Script,
 
+	// Used to track outpoint in the process of being claimed by our transactions. If height timer (u32) is expired
```
[the remark this answers is not in the document]
Comment (ariard) [reply src/ln/channelmonitor.rs]: Updated

File: src/ln/channelmonitor.rs [hunk 16]
```diff
@@ -1670,15 +1814,17 @@ impl ChannelMonitor {
 
 						htlc_timeout_tx.input[0].witness.push(Vec::new()); // First is the multisig dummy
 
-						htlc_timeout_tx.input[0].witness.push(their_sig.serialize_der().to_vec());
+						htlc_timeout_tx.input[0].witness.push(their_sig.clone().serialize_der().to_vec());
```
[the remark this answers is not in the document]
Comment (ariard) [reply src/ln/channelmonitor.rs]: My bad, was needed before I remove tracking for timeout transactions I guess

File: src/ln/channelmonitor.rs [hunk 17]
```diff
@@ -1670,15 +1814,17 @@ impl ChannelMonitor {
...
+						htlc_timeout_tx.input[0].witness.push(our_sig.clone().serialize_der().to_vec());
 						htlc_timeout_tx.input[0].witness[2].push(SigHashType::All as u8);
 
 						htlc_timeout_tx.input[0].witness.push(Vec::new());
-						htlc_timeout_tx.input[0].witness.push(chan_utils::get_htlc_redeemscript_with_explicit_keys(htlc, &local_tx.a_htlc_key, &local_tx.b_htlc_key, &local_tx.revocation_key).into_bytes());
+						let htlc_script = chan_utils::get_htlc_redeemscript_with_explicit_keys(htlc, &local_tx.a_htlc_key, &local_tx.b_htlc_key, &local_tx.revocation_key);
+						htlc_timeout_tx.input[0].witness.push(htlc_script.clone().into_bytes());
 
 						add_dynamic_output!(htlc_timeout_tx, 0);
+						//TODO: we should delay broadcast of HTLC-timeout tx until locktime - 1 to avoid being bounce off by mempool
```
[the remark this answers is not in the document]
Comment (ariard) [reply src/ln/channelmonitor.rs]: Going through core mempool rules the other day, I read that locktime transactions are going to be rejected from the mempool is locktime isn't equal or more to tip + 1 to avoid DoS vectors which means if we should broadcast timeout txn only at this given height to avoid having to rebuild it few blocks later. And for so track them with an OnchainEvent. It is how the mempool is working ?

File: src/ln/channelmonitor.rs (hunk 12 above)
Comment (ariard) [reply src/ln/channelmonitor.rs]: Yeah, was broken but not spotted by any other tests, seems it's really an edge case because if it's spending the funding output, dust htlcs are going to be cancelled on the behalf of remote commitment. Nevermind, switched the logic to be sure we have a previous_commitment or latest_commitment. Added  test_no_failure_dust_htlc_local_commitment, turn to true the is_local_tx in check_spend_local_transaction

File: src/ln/channelmonitor.rs (hunk 13 above)
Comment (ariard) [reply src/ln/channelmonitor.rs]: Updated

File: src/ln/channelmonitor.rs [hunk 18]
```diff
@@ -1234,8 +1345,10 @@ impl ChannelMonitor {
 							let predicted_weight = single_htlc_tx.get_weight() + Self::get_witnesses_weight(&[if htlc.offered { InputDescriptors::RevokedOfferedHTLC } else { InputDescriptors::RevokedReceivedHTLC }]);
 							if subtract_high_prio_fee!(self, fee_estimator, single_htlc_tx.output[0].value, predicted_weight, tx.txid()) {
 								let sighash_parts = bip143::SighashComponents::new(&single_htlc_tx);
-								sign_input!(sighash_parts, single_htlc_tx.input[0], Some(idx), htlc.amount_msat / 1000);
+								let (redeemscript, revocation_key) = sign_input!(sighash_parts, single_htlc_tx.input[0], Some(idx), htlc.amount_msat / 1000);
 								assert!(predicted_weight >= single_htlc_tx.get_weight());
+								assert_eq!(single_htlc_tx.input.len(), 1);
+								self.our_claim_txn_waiting_first_conf.insert(single_htlc_tx.input[0].previous_output.clone(), (height + 3, TxMaterial::Revoked { script: redeemscript, pubkey: Some(revocation_pubkey), key: revocation_key, is_htlc: true, amount: htlc.amount_msat / 1000 }, fee_estimator.get_est_sat_per_1000_weight(ConfirmationTarget::HighPriority)));
```
[the remark this answers is not in the document]
Comment (ariard) [reply src/ln/channelmonitor.rs]: On feerate, updated to cache only actual value used. In case we got block twice, I kept only the first version seen, doesn't seem efficient, in case of bump needed, to send two bumped transactions at same height even their respective previous states has different feerate base.

[the remark this answers is not in the document]
Comment (ariard) [reply src/ln/channelmonitor.rs]: And you may have the tx_broadcaster discard duplicate claim tx based on spend outpoint.

**Comment by ariard** [comment]
Updated at 0d3575a with your last review, I think we need to think more how we handle timelocks given https://github.com/rust-bitcoin/rust-lightning/pull/336#discussion_r300507662, https://github.com/rust-bitcoin/rust-lightning/pull/336#discussion_r300505061 and https://github.com/rust-bitcoin/rust-lightning/pull/336#discussion_r301756247

File: src/ln/channelmanager.rs [hunk 19]
```diff
@@ -2429,20 +2434,24 @@ impl ChainListener for ChannelManager {
...
-								// It looks like our counterparty went on-chain. We go ahead and
-								// broadcast our latest local state as well here, just in case its
-								// some kind of SPV attack, though we expect these to be dropped.
-								failed_channels.push(channel.force_shutdown());
-								if let Ok(update) = self.get_channel_update(&channel) {
-									pending_msg_events.push(events::MessageSendEvent::BroadcastChannelUpdate {
-										msg: update
-									});
+								log_trace!(self, "Detected channel-closing tx {} spending {}:{}, waiting until {} to close channel {}", tx.txid(), inp.previous_output.txid, inp.previous_output.vout, height + HTLC_FAIL_ANTI_REORG_DELAY - 1, log_bytes!(channel_id[..]));
+								match channel_closing_lock.entry(height + HTLC_FAIL_ANTI_REORG_DELAY - 1) {
```
[the remark this answers is not in the document]
Comment (ariard) [reply src/ln/channelmanager.rs]: Not something to do now, but I'm more and more thinking that channel_manager shouldn't be a ChainListener and get all blockchain events through ChannelMonitor and register outpoint to watch if needed. It would avoid inconsistencies between both of them due to different handling of block height timer.

File: src/ln/channelmanager.rs [hunk 20]
```diff
@@ -262,6 +262,12 @@ pub(super) struct ChannelHolder {
 	/// Messages to send to peers - pushed to in the same lock that they are generated in (except
 	/// for broadcast messages, where ordering isn't as strict).
 	pub(super) pending_msg_events: Vec<events::MessageSendEvent>,
+	/// Due to reorgs, we want to be sure that commitment txn has enough depth
+	/// (HTLC_FAIL_ANTI_REORG_DELAY).
+	/// before to close channel. Doing it to early would block us to pass backward
+	/// preimage in case of successful claim on onward channel. Key is block
+	/// height timer, value is array of channels id, waiting their closing.
+	pub(super) channel_closing_waiting_threshold_conf: HashMap<u32, Vec<[u8; 32]>>,
```
Comment (TheBlueMatt) [inline src/ln/channelmanager.rs]: Wait, why? If our counterparty broadcasts a commitment transaction, and it gets reorged out, we shoulnd't try to relay payments over the channel, that commitment transaction is probably gonna be mined again in a block soon, we should just move forward happily with a closed channel.

File: src/ln/channelmonitor.rs (hunk 17 above)
[the remark this answers is not in the document]
Comment (TheBlueMatt) [reply src/ln/channelmonitor.rs]: Ohh, I see your point. Still, broadcast_transaction *should* imply re-announcement (potentially at the appropriate time), if only because mempool transactions may time out. The docs on broadcast_transaction should make sure this is clear, but we shouldn't bother dealing with that.

File: src/ln/channelmanager.rs (hunk 19 above)
[the remark this answers is not in the document]
Comment (TheBlueMatt) [reply src/ln/channelmanager.rs]: Hmmm, see comment at https://github.com/rust-bitcoin/rust-lightning/pull/336/files#r304483221 I'd agree if block_connected in ChannelManager grows, but as long as we can keep it small, I don't see why its necessary.

File: src/ln/channelmanager.rs (hunk 20 above)
Comment (ariard) [reply src/ln/channelmanager.rs]: Oh I'm dumb, that's exactly what my test was trying to do, i.e testing some edge cases which is doable but doesn't make sense on a high-level. And from then, adding all this non-sense struct in channel_manager, I've dropped this commit and rewrote test_sweep_outbound_htlc_faiilure_update.

File: src/ln/channelmonitor.rs [hunk 21]
```diff
@@ -1883,10 +2173,40 @@ impl ChannelMonitor {
...
 		(watch_outputs, spendable_outputs, htlc_updated)
 	}
 
+	fn block_disconnected(&mut self, height: u32, block_hash: &Sha256dHash) {
+		if let Some(_) = self.onchain_events_waiting_threshold_conf.remove(&(height + ANTI_REORG_DELAY - 1)) {
+			//We may discard:
+			//- htlc update there as failure-trigger tx (revoked commitment tx, non-revoked commitment tx, HTLC-timeout tx) has been disconnected
+			//- our claim tx on a commitment tx output
+		}
+		self.our_claim_txn_waiting_first_conf.retain(|_, ref mut v| if v.0 == height + 3 { false } else { true });
```
Comment (TheBlueMatt) [inline src/ln/channelmonitor.rs]: Isn't this now wrong given get_height_timer? Maybe also track what block we saw the ting which is being spent explicitly to avoid any confusion.

File: src/ln/channelmanager.rs [hunk 22]
```diff
@@ -28,7 +28,7 @@ use secp256k1;
 use chain::chaininterface::{BroadcasterInterface,ChainListener,ChainWatchInterface,FeeEstimator};
 use chain::transaction::OutPoint;
 use ln::channel::{Channel, ChannelError};
-use ln::channelmonitor::{ChannelMonitor, ChannelMonitorUpdateErr, ManyChannelMonitor, CLTV_CLAIM_BUFFER, HTLC_FAIL_TIMEOUT_BLOCKS, HTLC_FAIL_ANTI_REORG_DELAY};
+use ln::channelmonitor::{ChannelMonitor, ChannelMonitorUpdateErr, ManyChannelMonitor, CLTV_CLAIM_BUFFER, GRACE_PERIOD_BLOCKS, ANTI_REORG_DELAY};
```
Comment (TheBlueMatt) [inline src/ln/channelmanager.rs]: New names are nice, but GRACE_PERIOD_BLOCKS isnt sufficiently clear, IMO, maybe NETWORK_SYNC_GRACE_PERIOD_BLOCKS?

File: src/ln/channelmonitor.rs [hunk 23]
```diff
@@ -1286,18 +1443,23 @@ impl ChannelMonitor {
...
-			let mut values_drain = values.drain(..);
 			let sighash_parts = bip143::SighashComponents::new(&spend_tx);
 
-			for (input, htlc_idx) in spend_tx.input.iter_mut().zip(htlc_idxs.iter()) {
-				let value = values_drain.next().unwrap();
-				sign_input!(sighash_parts, input, htlc_idx, value);
+			for (input, info) in spend_tx.input.iter_mut().zip(inputs_info.iter()) {
+				let (redeemscript, revocation_key) = sign_input!(sighash_parts, input, info.0, info.1);
+				let height_timer = Self::get_height_timer(height, info.2);
+				match self.our_claim_txn_waiting_first_conf.entry(input.previous_output.clone()) {
```
Comment (TheBlueMatt) [inline src/ln/channelmonitor.rs]: Doesn't need to happen in this PR, but in #347 this is gonna bite us - we don't track that all of these inputs are spent in the same transaction, with different expiries, so we may try to bump only one input, and then end up getting that confirmed, leaving nothing in the mempool trying to punish other inputs.
